modules/gdscript/gdscript_analyzer.cpp

```diff
@@ -615,431 +646,560 @@ GDScriptParser::DataType GDScriptAnalyzer::resolve_datatype(GDScriptParser::Type
+				signal_type.is_meta_type = false;
+				p_member.signal->parameters[j]->set_datatype(signal_type);
+			}
 
-				for (int j = 0; j < member.signal->parameters.size(); j++) {
-					GDScriptParser::DataType signal_type = resolve_datatype(member.signal->parameters[j]->datatype_specifier);
-					signal_type.is_meta_type = false;
-					member.signal->parameters[j]->set_datatype(signal_type);
-				}
-				// TODO: Make MethodInfo from signal.
-				GDScriptParser::DataType signal_type;
-				signal_type.type_source = GDScriptParser::DataType::ANNOTATED_EXPLICIT;
-				signal_type.kind = GDScriptParser::DataType::BUILTIN;
-				signal_type.builtin_type = Variant::SIGNAL;
+			// TODO: Make MethodInfo from signal.
+			datatype.type_source = GDScriptParser::DataType::ANNOTATED_EXPLICIT;
+			datatype.kind = GDScriptParser::DataType::BUILTIN;
+			datatype.builtin_type = Variant::SIGNAL;
 
-				member.signal->set_datatype(signal_type);
+			p_member.signal->set_datatype(datatype);
 
-				// Apply annotations.
-				for (GDScriptParser::AnnotationNode *&E : member.signal->annotations) {
-					E->apply(parser, member.signal);
-				}
-			} break;
-			case GDScriptParser::ClassNode::Member::ENUM: {
-				check_class_member_name_conflict(p_class, member.m_enum->identifier->name, member.m_enum);
-
-				GDScriptParser::DataType enum_type;
-				enum_type.type_source = GDScriptParser::DataType::ANNOTATED_EXPLICIT;
-				enum_type.kind = GDScriptParser::DataType::ENUM;
-				enum_type.builtin_type = Variant::DICTIONARY;
-				enum_type.enum_type = member.m_enum->identifier->name;
-				enum_type.native_type = p_class->fqcn + "." + member.m_enum->identifier->name;
-				enum_type.is_meta_type = true;
-				enum_type.is_constant = true;
-
-				// Enums can't be nested, so we can safely override this.
-				current_enum = member.m_enum;
-
-				for (int j = 0; j < member.m_enum->values.size(); j++) {
-					GDScriptParser::EnumNode::Value &element = member.m_enum->values.write[j];
-
-					if (element.custom_value) {
-						reduce_expression(element.custom_value);
-						if (!element.custom_value->is_constant) {
-							push_error(R"(Enum values must be constant.)", element.custom_value);
-						} else if (element.custom_value->reduced_value.get_type() != Variant::INT) {
-							push_error(R"(Enum values must be integers.)", element.custom_value);
-						} else {
-							element.value = element.custom_value->reduced_value;
-							element.resolved = true;
-						}
+			// Apply annotations.
+			for (GDScriptParser::AnnotationNode *&E : p_member.signal->annotations) {
+				E->apply(parser, p_member.signal);
+			}
+		} break;
+		case GDScriptParser::ClassNode::Member::ENUM: {
+			check_class_member_name_conflict(p_class, p_member.m_enum->identifier->name, p_member.m_enum);
+
+			GDScriptParser::DataType datatype;
+			datatype.kind = GDScriptParser::DataType::RESOLVING;
+			p_member.m_enum->set_datatype(datatype);
```

Comment:
Like you had a `bad_type` elsewhere, could you also just have a single `resolving_type` at the top of this function that gets set everywhere? This avoids having to `datatype = DataType()` to reuse all the time?

Reply:
good point, thats much cleaner